read-fonts/src/collections/range_set.rs

```diff
@@ -0,0 +1,395 @@
+//! Stores a disjoint collection of ranges over numeric types.
+//!
+//! Overlapping and adjacent ranges are automatically merged together.
+
+use core::{
+    cmp::{max, min},
+    fmt::{Debug, Formatter},
+    iter::Peekable,
+    ops::RangeInclusive,
+};
+use std::collections::BTreeMap;
+
+use types::Fixed;
+
+#[derive(Default, Clone, PartialEq, Eq)]
+/// A set of disjoint ranges over numeric types.
+///
+/// Overlapping and adjacent ranges are automatically merged together.
+pub struct RangeSet<T> {
+    // an entry in the map ranges[a] = b implies there is an range [a, b] (inclusive) in this set.
+    ranges: BTreeMap<T, T>,
+}
+
+/// Allows a two values to be tested for adjacency.
+pub trait OrdAdjacency<T> {
+    /// Returns true if self is adjacent to rhs
+    fn are_adjacent(self, rhs: T) -> bool;
```

Comment:
we should clarify if sidedness matters, or if it is strictly about testing increasing order.